### ISSUE TITLE
Bug: [strict-boolean-expressions] fails with eslint v9.39.0

### ISSUE THREAD

**Issue by larsmunkholm** [typescript-eslint/typescript-eslint#11734]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://bolt.new/~/sb1-fsvbwljw

### Repro Code

```TypeScript
{error != null && <Alert severity="error">{t(error)}</Alert>}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/strict-boolean-expressions": [
       "error",
       { allowNumber: false, allowString: false, allowNullableBoolean: true },
    ]
  },
};
```

### tsconfig

```jsonc

```

### Expected Result

No errors

### Actual Result

1200+ errors

### Additional Info

After the update to v9.39.0, I get 1200+ linting errors in my project, on all React components rendered with a condition shorthand like this

```js
{error != null && <Alert severity="error">{t(error)}</Alert>}
```

Downgrading eslint to v9.38.0 helps.
I don't know if this is for you or them to fix, so I have created an issue with them as well here: https://github.com/eslint/eslint/issues/20277

**Comment by ritschwumm** [comment]
thanks @larsmunkholm, i've run into the same problem but couldn't bring myself to make a reproducer.

**Comment by bradzacher** [comment]
There is a new feature that eslint shipped in v9.39.0 that is breaking many rules.

We have asked them to revert it in https://github.com/eslint/eslint/issues/20272

**Comment by bradzacher** [comment]
https://eslint.org/blog/2025/11/eslint-v9.39.1-released/

The new version of eslint has been released which should resolve this.
